File: app/src/main/java/com/jtl/arruler/helper/SessionHelper.java
package com.jtl.arruler.helper;

import android.content.Context;
import android.util.Log;

import com.google.ar.core.Session;
import com.google.ar.core.exceptions.CameraNotAvailableException;
import com.google.ar.core.exceptions.UnavailableApkTooOldException;
import com.google.ar.core.exceptions.UnavailableArcoreNotInstalledException;
import com.google.ar.core.exceptions.UnavailableDeviceNotCompatibleException;
import com.google.ar.core.exceptions.UnavailableSdkTooOldException;
import com.google.ar.core.exceptions.UnavailableUserDeclinedInstallationException;
import com.socks.library.KLog;

/**
 * 作者:jtl
 * 日期:Created in 2019/5/2 17:01
 * 描述:Session单例类
 * 更改:
 */
public class SessionHelper {
    private static String TAG = SessionHelper.class.getSimpleName();
    private Session mSession;

    private SessionHelper() {
    }

    public static SessionHelper getInstance() {
        return SessionHelperHolder.sSessionHelper;
    }

    public void initSession(Context context) {
        Exception exception = new Exception();
        String message = "";
        try {
            mSession = new Session(context);
            mSession.resume();
        } catch (UnavailableDeviceNotCompatibleException e) {
            exception = e;
            message = "This device does not support AR";
        } catch (UnavailableArcoreNotInstalledException e) {
            exception = e;
            message = "Please install ARCore";
        } catch (UnavailableSdkTooOldException e) {
            exception = e;
            message = "Please update ARCore";
        } catch (UnavailableApkTooOldException e) {
            exception = e;
            message = "Please update this app";
        } catch (CameraNotAvailableException e) {
            exception = e;
            message = "Session resume failed";
        }catch (Exception e){
            exception = e;
            message = e.toString();
        }
        KLog.e(exception + ":" + message);
    }

    public Session getSession(Context context) {
        if (mSession == null) {
            initSession(context);
        }

        return mSession;
    }

    public void resumeSession() {
        try {
            if (mSession != null) {
                mSession.resume();
            }
        } catch (CameraNotAvailableException e) {
            Log.e(TAG, "Camera not available. Please restart the app.");
            mSession = null;
        }
    }

    public void pauseSession() {
        if (mSession != null) {
            mSession.pause();
        }
    }

    public void closeSession(){
        if (mSession != null) {
            mSession.close();
        }
    }

    private static class SessionHelperHolder {
        private final static SessionHelper sSessionHelper = new SessionHelper();
    }
}

File: app/src/main/java/com/android/ar_ruler_kt/helper/SessionHelper.kt
package com.android.ar_ruler_kt.helper

import android.app.Activity
import android.content.Context
import android.util.Log
import com.google.ar.core.*
import com.google.ar.core.exceptions.*
import java.util.*

/**
 * @author：TianLong
 * @date：2022/6/27 20:59
 * @detail：ARCore Session Helper类
 */
object SessionHelper:HelperInterface{
    val featureSet: Set<Session.Feature> = setOf()
    lateinit var session: Session

    private val sessionConfig: Config by lazy {
        return@lazy Config(session).setFocusMode(Config.FocusMode.AUTO)
            .setUpdateMode(Config.UpdateMode.BLOCKING)
            .setPlaneFindingMode(Config.PlaneFindingMode.HORIZONTAL_AND_VERTICAL)
    }
    fun initialize(context:Context):Boolean{
        return if (prepare(context)){
            try {
                session = Session(context,featureSet)

                val cameraConfigFilter = CameraConfigFilter(session)
                val list = session.getSupportedCameraConfigs(cameraConfigFilter)
                cameraConfigFilter.targetFps = EnumSet.of(CameraConfig.TargetFps.TARGET_FPS_60)
                cameraConfigFilter.depthSensorUsage = EnumSet.of(CameraConfig.DepthSensorUsage.DO_NOT_USE)
                for (con in list){
                    Log.w(TAG,"id:${con.cameraId}   width:${con.imageSize.width}  height:${con.imageSize.height} upper:${con.fpsRange.upper }   lower${con.fpsRange.lower}")
                }
                session.cameraConfig=list.first()


                session.configure(session.config.apply {
                    lightEstimationMode = Config.LightEstimationMode.ENVIRONMENTAL_HDR

                    // Depth API is used if it is configured in Hello AR's settings.
                    depthMode =
                        if (session.isDepthModeSupported(Config.DepthMode.AUTOMATIC)) {
                            Config.DepthMode.AUTOMATIC
                        } else {
                            Config.DepthMode.DISABLED
                        }
                    depthMode =  Config.DepthMode.DISABLED
                    focusMode = Config.FocusMode.AUTO
                    updateMode = Config.UpdateMode.LATEST_CAMERA_IMAGE
                    planeFindingMode = Config.PlaneFindingMode.HORIZONTAL_AND_VERTICAL
                })
            }catch (exception:Exception){
                var msg = when (exception) {
                    is UnavailableUserDeclinedInstallationException ->
                        "Please install Google Play Services for AR"
                    is UnavailableApkTooOldException -> "Please update ARCore"
                    is UnavailableSdkTooOldException -> "Please update this app"
                    is UnavailableDeviceNotCompatibleException -> "This device does not support AR"
                    is CameraNotAvailableException -> "Camera not available. Try restarting the app."
                    else -> "Failed to create AR session: $exception"
                }
                Log.e(TAG,msg)
                return false
            }

            true
        }else{
            false
        }
    }

    fun release(){
        session.close()
    }

    private fun prepare (context: Context):Boolean{
        when {
            (!CameraPermissionHelper.hasCameraPermission(context as Activity))-> {
                CameraPermissionHelper.requestCameraPermission(context)
                Log.w(TAG,"ARCore must have camera permission")
                return false
            }

            ArCoreApk.getInstance().requestInstall(context,true) == ArCoreApk.InstallStatus.INSTALL_REQUESTED ->{
                Log.w(TAG,"ARCore must install ArCoreApk")
                return false
            }
        }
        Log.w(TAG,"ARCore has prepared")
        return true
    }
}